Update requirements files, use pyproject.toml, update readme.

wylie102 opened the pull request:
This commit adds a pyproject.toml for managing dependencies and dev tool configurations.

Dev dependancies are updated slightly:
- Ruff replaces black, but is configured to format in almost exactly the same way.
- The script dependencies such as beautifulsoup are added to the dev group.
- Codespell is added as a tool, this catches common spelling mistakes using a dictionary of common errors (so it won't highlight every little issue in function/variable names etc). There is a list of ignored words in it's config in the pyproject.toml file, feel free to add to it. Run it using condespell <path/to/files>.

The requirements files are updated to specify the current versions of duckdb and the dbt connectors, for those wanting to stick with pip.
Also the common.in is renamed to dev.in and we have a dev.txt file. The duckdb.txt and postgres.txt no longer include the dev dependencies.

The readme file is updated to show uv vs pip instructions.

katy-sadowski commented on README.md:
```diff
@@ -19,13 +19,138 @@ By default, the project will source the Synthea and OMOP vocabulary data from se
+> Note: The --group and --dev arguments must be passed at the same time, passing them separately e.g `uv sync --group duckdb` `uv sync --dev` will just result in the last group specified being installed.
+
+ - If you are using VS Code, create a .env file in  the root of your repo workspace (`touch .env`) and add a PYTHONPATH entry for your virtual env (for example, if you cloned your repo in your computer's home directory, the entry will read as: `PYTHONPATH="~/dbt-synthea/.venv/bin/python"`)
+ - Now, in VS Code, once you set this virtualenv as your preferred interpreter for the project, the vscode config in the repo will automatically source this env each time you open a new terminal in the project.  Otherwise, each time you open a new terminal to use dbt for this project, run:
```
this is not working for me anymore. i thought it might be because this line was referencing `dbt-env` from the old instructions: https://github.com/OHDSI/dbt-synthea/blob/8fd7e0eb1c515c4da978e6448f7375a1c8c778b6/.vscode/settings.json#L7

however, even after updating to `.venv` it's still not working.  any ideas?

(ps, i realize the inclusion of this VS code stuff is a bit controversial but i really would like to make this bit seamless for folks less familiar with environment setup)

katy-sadowski replied:
OK wow, the hack suggested in the description of https://github.com/astral-sh/uv/issues/9637 works!  I updated the default interpreter in .vscode/settings.json as follows:

`"python.defaultInterpreterPath": "./.venv",`

The first time you are in VS Code, you either need to enter the full path manually OR close (not reload) the window and reopen.  Then subsequent times, the env activates automatically.

wylie102 replied:
> OK wow, the hack suggested in the description of [astral-sh/uv#9637](https://github.com/astral-sh/uv/issues/9637) works! I updated the default interpreter in .vscode/settings.json as follows:
> 
> `"python.defaultInterpreterPath": "./.venv",`

When I did this I found that using `./.venv/bin` worked for me but not `./.venv`. Could you check and see if that works for you because i'd be tempted to include that.

katy-sadowski replied:
Sorry I missed this!!!  It works for me with ./.venv/bin as well (and out-of-the-box with the updated file in .vscode).

I might just add instructions for (1) activating the venv for the first time, and (2) subsequent times (i.e., close VS Code and reopen - this is the only way I've gotten the auto-activate to work. Reloading the window does not work).  Btw, pretty sure these same instructions would apply to the repo in its current state, they were just missing before :)

Thanks!

katy-sadowski commented on README.md:
````diff
@@ -19,13 +19,138 @@ By default, the project will source the Synthea and OMOP vocabulary data from se
+ uv sync --group duckdb --dev
+ ```
+
+> Note: The --group and --dev arguments must be passed at the same time, passing them separately e.g `uv sync --group duckdb` `uv sync --dev` will just result in the last group specified being installed.
````
is it possible to change this so we can install dependencies for multiple databases?  i would like to have both duckdb and postgres requirements installed.

katy-sadowski reviewed (commented):
Looking good!  What are your thoughts on actually ditching the pip instructions altogether and only supporting uv?  I think it's a bit confusing to have multiple ways to do the same thing.  If we provide a bit more guidance on the uv installation, I think it should be OK, unless there's some downside you know of?

Edit: Forgot to mention, the setup instructions with uv work for me on a fresh clone of the repo :)

wylie102 commented:
> What are your thoughts on actually ditching the pip instructions altogether and only supporting uv? 

I don't mind it, I consider the pip instructions kind of legacy. Especially for projects like this where you don't want to install the project itself.

>I think it's a bit confusing to have multiple ways to do the same thing. 

This is kind of why I relegated the entire pip section to the bottom of the readme. But I can understand it's a bit cluttered and repetative.

>Unless there's some downside you know of?

I was wondering if someone might be working on a device where they can't install uv for some reason. Although then I guess `pipx install uv` or `pip install uv` would still work.

It's up to you, I started using uv very shortly after starting to learn python so I've barely used pip. But there might be some who are more tied to it? I'm fine to drop it entirely though, it would let us get rid of the requirements files and just have everything work from the pyproject.toml. 

Or we could move the pip instructions to a separate .md file and put that in with the requirements? I'm pretty sure we can get the requirements to be refreshed automatically by uv using pre-comit, although we'd have to remember to update the .md on a change. I'm sure there must be some tools for that too though, even just a regex would do it to turn `uv run` into `python3`.

But probably dropping it entirely is cleaner. We'd just have to update some of the github actions to use uv. But that shouldn't be an issue.

wylie102 commented:
> Its probably dropping it entirely, but the reason I never advised / brought in uv - is precisely as its quite an opinionated way of doing it, and makes it harder for new people to use it (even if uv is objectively better)

Yeah, this is a much more succinct way of putting it 😂. 

And practically, if they are very new to everything and are just copy pasting the commands then asking them to run `curl -LsSf https://astral.sh/uv/install.sh | sh` or `pip install uv` and then the rest will make no less sense to them and not be any more difficult.

And if they're some curmudgeonly old dev who refuses to move on from pip then they probably know enough about pip to just copy paste the groups from the .toml into a their own requirements.in and go from there.

So it probably won't cause any issues with anyone just dropping it entirely.

katy-sadowski commented:
> And practically, if the user is very new to everything and are just copy pasting the commands, then asking them to run curl -LsSf https://astral.sh/uv/install.sh | sh or pip install uv before the other commands will make no less sense to them and not be any more difficult.
> And if they're some curmudgeonly old dev who refuses to move on from pip then they probably know enough about pip to just copy paste the groups from the .toml into a their own requirements.in and go from there.
> So it probably won't cause any issues with anyone just dropping it entirely.

This is my thinking as well.  Let's go full uv!  @wylie102 if you don't mind adding that uv install step as a note, that'd be great!

wylie102 commented:
Okay after a bit of trial and error I have added a test workflow. Unlike the docs one this should run on any push, so if you push on any branch. 

So if you push a change to your own branch it should run - dbt build, dbt test, and dbt docs generate - using uv and the duckdb requirements. 

I've updated the generate docs to use the same methods so it should run (although it hasn't been triggered in my fork as I haven't pushed these to main.

I've also updated the readme.

The results of the tests aren't currently persisted, possibly we could add another part to run this on a pull request and include the results in it in a later PR. We could also add a run of the formatting tools etc if we wanted. Or possibly the recce tool that lawrence mentioned (or a ducklake snapshot if we are feeling adventurous)  But it's probably best just to check that this works correctly in a simple version first.

Take a look and let me know if there are any changes to that or to the readme that need to be made.

lawrenceadams commented on .github/workflows/test-dbt-duckdb.yml:
```diff
@@ -0,0 +1,58 @@
+name: Test dbt duckdb
+
+on:
+  push:
+    branches:
+      - '**'
+    paths:
+      - macros/**
+      - models/**
+      - scripts/**
+      - seeds/**
+      - snapshots/**
+      - tests/**
+      - dbt_project.yml
+      - .github/workflows/**
+  workflow_dispatch:
```
I'd get rid of workflow dispatch and change it to run on PR instead of branches, otherwise you're gonna needlessly burn through GH compute credits

lawrenceadams replied:
```yaml
on:
  pull_request:
    branches: [main]
    paths:
      - models/**
      - seeds/**
      - snapshots/**
      - tests/**
      - macros/**
      - dbt_project.yml
      - packages.yml
      - requirements.txt
    types:
      - opened
      - synchronize
      - reopened
```

This is what I tend to go for - only runs when there's a PR open for something

Unsure but @katy-sadowski may need to change a setting so that only maintainers can authorise workflow runs

wylie102 replied:
Oh I literally didn't know about the gh compute credits, yeah PR makes more sense. That's also why I hadn't removed the documents build, I didn't think there was really a down side to checking it worked (even though I couldn't really think of a case that it wouldn't if run and test did). But I'll change both of those.

I mostly was setting this up to run on a contributors fork, would that then come from their credits quota then? Do you k ow if we move it to or does it charge the ohdsi repository or the individual forks?

katy-sadowski replied:
The proposal to run actions on PRs to main sounds good to me!  And I've never used/set up workflow dispatch before, but I think it's OK to have it as an option in the repo.  It should only be available to people with write access to the repo.

GH Actions runs in forks will be charged to the user who owns the fork, and need to be set up in the fork before they'll run automatically.

lawrenceadams commented on pyproject.toml:
```diff
@@ -0,0 +1,123 @@
+
+[dependency-groups]
+dev = [
+  "jinja2",
+  "pre-commit",
+  "sqlfluff",
+  "sqlfluff-templater-dbt",
+  "ruff",
+  "beautifulsoup4",
+  "ruamel-yaml",
+  "requests",
+  "codespell",
```
Are these all needed explicitly rather than as a known dependency? Why is jinja explicitly mentioned for example if it's a known dependency of dbt (for example)

wylie102 replied:
I just copied them across from common.in 

https://github.com/OHDSI/dbt-synthea/blob/e119df64bb3f99394b8d124e2f72a3798b343820/requirements/common.in#L1-L5

and added the script dependencies. If it's not being used independently from dbt (like with a linter/formatter extension or something) I'll take it out.

katy-sadowski replied:
I added Jinja2 because there was a Dependabot [alert](https://github.com/OHDSI/dbt-synthea/security/dependabot/1) for the package and this was the recommended fix (https://github.com/OHDSI/dbt-synthea/pull/122).

wylie102 replied:
Looking at the dbt requirements, they specify >=3.12, so it might be worth us still pinning it to >=3.15. It won't do any harm. I'll probably move it to the duckdb and Postgres groups though, as I haven't specified the version on any of the dev dependencies as I thought we could figure it out ourselves if there were any issues, and if we run into any big differences between the versions we can always add a version specification later.

lawrenceadams commented on .github/workflows/test-dbt-duckdb.yml:
```diff
@@ -0,0 +1,58 @@
+      - run: uv run dbt run
+      - run: uv run dbt test
+
+      - name: Generate DBT Docs
+        run: uv run dbt docs generate
```
Why do this here?

wylie102 replied:
Mostly just a hangover from me using your generate docs as a base, didn't see the harm in keeping it but I wasn't thinking about the compute credits.

lawrenceadams replied:
Ah yeah its more as its redundant in this instance as there is the `main` pipeline run

lawrenceadams reviewed (commented):
Mostly pretty good - as per comments if those make sense
